src/DotNetWorker.Core/Context/Features/DefaultFunctionInputBindingFeature.cs

```diff
@@ -10,31 +10,25 @@
 
 namespace Microsoft.Azure.Functions.Worker.Context.Features
 {
-    internal class DefaultModelBindingFeature : IModelBindingFeature
+    internal class DefaultFunctionInputBindingFeature : IFunctionInputBindingFeature
     {
-        private bool _inputBound;
-        private object?[]? _parameterValues;
+        private FunctionInputBindingResult? _inputBindingResult;
         private readonly IConverterContextFactory _converterContextFactory;
 
-        public DefaultModelBindingFeature(IConverterContextFactory converterContextFactory)
+        public DefaultFunctionInputBindingFeature(IConverterContextFactory converterContextFactory)
         {
             _converterContextFactory = converterContextFactory ??
                                        throw new ArgumentNullException(nameof(converterContextFactory));
         }
 
-        public object?[]? InputArguments => _parameterValues;
-
-        public async ValueTask<object?[]> BindFunctionInputAsync(FunctionContext context)
+        public async ValueTask<FunctionInputBindingResult> BindFunctionInputAsync(FunctionContext context)
         {
-            if (_inputBound)
+            if (_inputBindingResult is not null)
```

Comment:
We should synchronize access here to make sure we don't end up with races leading to this logic unexpectedly running multiple times as that could potentially lead to unpredictable behavior.

Reply:
The race was already possible here, so not new, but the changes do  increase the possibility of this happening.

Reply:
Good catch. SemaphoreSlimmed.